src/ConfigServiceProvider.php

```diff
@@ -68,169 +57,67 @@ public static function fromConfig(array $config, array $settings = [])
      */
     public static function fromFiles(array $patterns, array $settings = [])
     {
-        $locator = new FileLocator();
-        $files   = $locator->locate($patterns);
+        $separator = self::getSettingOrDefault(self::SETTING_SEPARATOR, $settings, self::DEFAULT_SEPARATOR);
 
-        if (empty($files)) {
-            throw new NoMatchingFilesException(
-                'No files found matching patterns: ' . implode(', ', $patterns)
-            );
-        }
-
-        $factory = new ReaderFactory([
-            '.json' => 'TomPHP\ConfigServiceProvider\JSONFileReader',
-            '.php'  => 'TomPHP\ConfigServiceProvider\PHPFileReader',
-        ]);
-
-        $configs = array_map(
-            function ($filename) use ($factory) {
-                $reader = $factory->create($filename);
-                return $reader->read($filename);
-            },
-            $files
-        );
-
-        $config = call_user_func_array('array_replace_recursive', $configs);
-
-        return self::fromConfig($config, $settings);
+        return self::fromConfig(ApplicationConfig::fromFiles($patterns, $separator), $settings);
     }
 
     /**
      * @api
      *
-     * @param array                         $config
-     * @param string                        $prefix
-     * @param string                        $separator
-     * @param ConfigurableServiceProvider[] $subProviders
+     * @param array|ApplicationConfig    $config
+     * @param string                     $prefix
+     * @param string                     $separator
+     * @param ServiceProviderInterface[] $subProviders
```

Comment:
@tomphp 

Seems like this argument doesn't actually exist anymore, should we remove it from the docblock?